### ISSUE TITLE
pmp-granularity detection assumes pmpcfg is reset to 0

### ISSUE THREAD

**Issue by neelgala** [riscv-software-src/opensbi#225]
The current method for checking pmp granularity is to write all ones to the pmpaddr0 and then checking the number of trailing zeros on a read of the same csr. However, this only works if the pmpcfg0.A as been configured to TOR or OFF. But reset values of pmpcfg.A is implementation defined. For example, Spike resets all the pmpcfg.A to NAPOT. Which means the lower order bits will all be read as ones instead of zeros.

Thus opensbi needs to first set pmpcfg0.A to 0 and then check for trailing zeros in pmpaddr0.

**Comment by avpatel** [comment]
Yes, we have been assuming that all PMPCFGx will be set to zero upon reset.

This is a good catch.

Can any of your folks send patch to mailing list for fixing this in hart_pmp_get_allowed_addr() ?

Regards,
Anup
